### DISCUSSION TITLE
How to use fetch or axios in Flow

### DISCUSSION THREAD

**Discussion by Lucas-Gruber** [directus/directus#17773] [Archive]
Hi!
I have a Next.JS project with an API route that is triggered every day by a flow with a Post Req. Unfortunately this takes too long to be used in Vercel.

Now I wanted to do it in such a way that I split this route into pages. That has worked so far.

**Now I come to my problem:**
But now this flow has to send post requests with a body to this API in a for-loop. The only way to do this would be in a run script opteration since I can't otherwise send web requests in a loop. (Feel free to correct me if I do) I've tried using the Fetch API first, but it always fails. Then I allowed "axios" via the "FLOWS_EXEC_ALLOWED_MODULES" variable. Then I tried the script with axios, unfortunately that didn't work either.

Unfortunately I don't know what else I can try. I am writing the script below. I hope someone can help me here.

**script:**
```js
module.exports = async function(data) {
	const test = await fetch("https://jsonplaceholder.typicode.com/todos/1").then((res) => {return res.json()})
	return await test;
}
```

**Comment by alexvdvalk** [comment]
You can use the "trigger flow" operation to loop through an array. Have you tried this?

https://docs.directus.io/cookbook/flows/flows-for-loop.html

**Comment by xu4wang** [comment]
it says https://docs.directus.io/configuration/flows/operations.html

```
.....
 This means, for example, no access to the file system and no ability to do network requests.
```

I guess it's not possible

**Comment by domdus** [comment]
**! Since [v.10.6.0](https://github.com/directus/directus/releases/tag/v10.6.0) this does not work with `Run Script` anymore.**
vm2 got replaced with isolated-vm. To still use Axios or other node modules in flows, you need to wrap them in a [custom operation](https://docs.directus.io/extensions/operations.html) extension.

--------------------------------------

1. Add axios to FLOWS_EXEC_ALLOWED_MODULES in env config
2. Restart directus, after that you should see the following message below the Flows - Run Script input field:
![run_script_axios](https://user-images.githubusercontent.com/67951828/227049957-c69bc9fe-91f3-42ab-80a3-4b0f6388b26e.jpg)

3. Use axios lib, example:

```
const axios = require('axios');

module.exports = async function({$trigger}) {
        // add your loop logic here?
        try {
            const response = await axios.post('https://domain.com',
                    {
                        someJsonBody: "text",
                        addDataFromTrigger: $trigger.payload.field
                    },
                    {
                        headers: {
                            Authorization: "xxx"
                        }
                    }
            );
    
            return response.data;
        }
        catch (e) {
            return { "exception": e.message};
        }
}
```

**Comment by rijkvanzanten** [comment]
Be very careful with making requests like that though. It circumvents security features like `IMPORT_IP_DENY_LIST` and could open up SSRF attack vectors

**Reply by notflip** [reply]
@rijkvanzanten What is the recommended approach to perform requests from a flow?

**Reply by rijkvanzanten** [reply]
Use the request operation

**Reply by igortsev** [reply]
Send request it’s good but very hard to work with response because you need to make console log to see it and add some additional operations. Can we run some way with debugging ?

**Comment by Blaxzter** [comment]
The work around with trigger another flow for "loops" is not really intuitive and I would wish for a different implementation.

**Comment by malipetek** [comment]
While migrating from vm2 to isolated-vm `FLOWS_EXEC_ALLOWED_MODULES ` was gone with the version [v10.6.0](https://github.com/directus/directus/releases/tag/v10.6.0)

**Comment by bernardcosta** [comment]
Also to mention any HTTP request with  built in pagination will not work with webhooks/request operation.
